### ISSUE TITLE
dropmissing returns boundserror

### ISSUE THREAD

**Issue by laashe** [JuliaData/DataFrames.jl#1838]
I have a DataFrame that was essentially initated by

```julia
julia> model = DataFrame(Matrix{Missing}(undef, numrows, numcols), myfieldnames::Vector{Symbol});
```
before converting column types and filling them from another DataFrame (that was read from file with CSV.read()). When I try running the following I get into trouble.
```julia
julia> typeof(model)
DataFrame
julia> dropmissing(model)
BoundsError: attempt to access 199-element Array{String,1} at index [210]

Stacktrace:
 [1] setindex! at ./array.jl:767 [inlined]
 [2] setindex! at ./multidimensional.jl:460 [inlined]
 [3] _invert(::Dict{String,UInt32}) at /Users/laashe/.julia/packages/PooledArrays/ufJSl/src/PooledArrays.jl:22
 [4] PooledArrays.PooledArray(::PooledArrays.RefArray{Array{UInt32,1}}, ::Dict{String,UInt32}) at /Users/laashe/.julia/packages/PooledArrays/ufJSl/src/PooledArrays.jl:36
 [5] copy(::CSV.Column{String,PooledString}) at /Users/laashe/.julia/packages/CSV/IwqOm/src/tables.jl:80
 [6] (::getfield(DataFrames, Symbol("##DataFrame#83#86")))(::Bool, ::Type, ::Array{AbstractArray{T,1} where T,1}, ::DataFrames.Index) at /Users/laashe/.julia/packages/DataFrames/CZrca/src/dataframe/dataframe.jl:127
 [7] Type at /Users/laashe/.julia/packages/DataFrames/CZrca/src/dataframe/dataframe.jl:97 [inlined]
 [8] getindex(::DataFrame, ::UnitRange{Int64}) at /Users/laashe/.julia/packages/DataFrames/CZrca/src/dataframe/dataframe.jl:302
 [9] completecases at /Users/laashe/.julia/packages/DataFrames/CZrca/src/abstractdataframe/abstractdataframe.jl:598 [inlined]
 [10] #dropmissing#49(::Bool, ::Function, ::DataFrame, ::UnitRange{Int64}) at /Users/laashe/.julia/packages/DataFrames/CZrca/src/abstractdataframe/abstractdataframe.jl:668
 [11] dropmissing at /Users/laashe/.julia/packages/DataFrames/CZrca/src/abstractdataframe/abstractdataframe.jl:668 [inlined] (repeats 2 times)
 [12] top-level scope at In[31]:1
```
However, for some reason, the following works.
```julia
julia> dropmissing(first(model, size(model, 1));
```
I'm running this in a jupyter notebook through IJulia, and encountered this issue only after upgrading from julia v1.1 to v1.1.1. No such issue on v1.1.

```julia
julia> versioninfo()
Julia Version 1.1.1
Commit 55e36cc (2019-05-16 04:10 UTC)
```

Any idea why this happens?

**Comment by bkamins** [comment]
@quinnj it seems to be a problem with `copy` in CSV.jl.

**Comment by quinnj** [comment]
can anyone share a full reproduction? A file to read w/ CSV.jl that then shows this problem? That would help immensely in solving this.

**Comment by bkamins** [comment]
@laashe I cannot reproduce the error either. In particular - are you sure that you are calling `dropmissing` on `model` data frame. The error message you report is only possible if you called `dropmissing` on a data frame read-in from a file (as @quinnj indicates) not a data frame manually created by converting a matrix (what you have indicated in your post above).

**Comment by laashe** [comment]
@quinnj @bkamins thanks for swift replies.

The data I am handling are sensitive, meaning I can't hand out the full set. In attempting to generate a subset I found an indication on there being a certain row causing the issue. Total file size is about 500000 rows, so I tried the following.
```
full set             => error
lines 1 - 300        => no error
"OK, let's be systematic"
1 - 400000 => error
1 - 300000 => error
1 - 200000 => error
1 - 100000 => no error
"Aha, it must be a line between 100000 and 200000"
100000 - 200000 => error
150000 - 200000 => no error
140000 - 200000 => error
"Aha again, it must be between 140000 and 150000
140000 - 150000 => no error
139999 - 150001 => no error
130000 - 160000 => no error
"???"
```
The subset was taken by simply using `head` and `tail`. The conclusion is that there doesn't seem to be a row that causes this, at least not a single one.  I am thus not able to create a subset of the file that reproduces the error while still being small enough to not be sensitive for the client. The next step would be to replace the data in the file with dummy data, and I tried that, leaving missing values as is, strings with "S", Int64s with `0`, Dates with `Dates.Date(2001)`, and writing back to a file, but reading that file doesn't reproduce the issue.

I am still experiencing the issue, and yes, I am certain I am calling dropmissing on the `model` DataFrame. Below is the code I run to reproduce the issue for myself.

```julia
#=
The main goal of the code is to read in a file of a certain format and amssage it inot a target format.
The raw data is referred to as "internal", while the target is "model".
The target model has a set of predefined fields, some of which are optional, some are required.
Some of the fields in the target format have an equivalent in the internal data.
=#

# enlarge preview display
ENV["LINES"]=200;
ENV["COLUMNS"]=2000;

# load packages
using DataFrames, CSV, Dates

# read internal file
internalfile = "../Skadestatistikk_20190501_utf8.txt"
internal = CSV.read(
    internalfile,
    delim = '\t'
);

# count rows
numrows = size(internal,1)

#=
Here I am reading in a file listing the potential fields in model
=#
modelfields = CSV.read(
    "../fields_juggling/model_except_always_derived.txt",
    delim = "\t",
    types = [String, String]
)

# parse field names to symbols
modelfields[:field] = Symbol.(modelfields[:field]);

#=
Some model fields have an equivalent in the internal data, here I create a dictionary from a list of those
=#
dictmodel2internal = CSV.read(
    "../fields_juggling/dict_model_to_internal_fields.txt",
    delim = '\t'
)
dictmodel2internal = Dict(zip(dictmodel2internal[:model], dictmodel2internal[:internal]));

#=
Here I initiate the empty output model dataframe.
I use Missing because I don't want to deal with #undef.
=#
numcols = size(modelfields, 1)
model = DataFrame(
    Matrix{Missing}(undef, numrows, numcols),
    modelfields[:field]
)

# Here I populate model fields with equivalent internal fields
map(
    k -> (
        model[Symbol(k)] = internal[Symbol(dictmodel2internal[k])]
    ),
    collect(keys(dictmodel2internal))
);
internal = [];

#=
Here I read a list of required model fields, and create an intersect with those that should have data,
i.e. the ones that had an equivalent in the internal data.
=#
requiredfields = CSV.read(
    "../fields_juggling/required_model_fields.txt",
    delim = '\t'
)

#=
This call to dropmissing fails. typeof(model) returns DataFrame.
If I use first(model, size(model, 1)) instead it succeeds.
=#
model = dropmissing(
    model,
    Symbol.(intersect(requiredfields[:field], collect(keys(dictmodel2internal))))
);
numrows = size(model, 1)
```

So in short I read a tsv file, initiate a DataFrame with missing values, and fill the latter with certain columns of the former, before I try to drop the missing rows of the filled model columns. I am not able to make a non-sensitive subset that reproduces the error, and since it goes away just by using the `first()` command I guess there is not much need for action here. If you have some ideas of what I could do to further narrow down the issue I'd love to contribute.

**Comment by bkamins** [comment]
Thank you for a detailed report! Now I see that you move columns from another data frame to `model` before `dropmissing`. I will try to investigate this.

**Comment by bkamins** [comment]
@quinnj If I read the error message correctly this should be only possible if for some reason refs for some column contain duplicates. Are refs modified anywhere so that this could happen?

**Comment by laashe** [comment]
I'm not sure what you are referring to, not being an expert in Julia, but with respect to duplicates two of the columns in `model` are filled from the same column in `internal`. I don't know if that information is relevant, but now you have it.

**Comment by bkamins** [comment]
@laashe I was referring to how `copy` in CSV.jl internally is implemented. But thank you for the remark.

**Comment by quinnj** [comment]
Sorry, I just remember this issue; is it still relevant? Did we ever boil things down to a reproduction that exhibits a problem? I couldn't determine if there's something wrong from this discussion or not.

**Comment by bkamins** [comment]
There is (or was) something wrong, but I was never able to reproduce what @laashe reported.

**Comment by laashe** [comment]
Since the issue, whatever it is, is avoided by using `first()`, I would say spend your time on more pressing matters. Not able to make a reliable reproduction of the issue either, but thank you for your time!

**Comment by bkamins** [comment]
Feel free to reopen the issue when you stumble on the problem again. Thank you!
